Modules/Effect/Effect/IDiscoverCollectionType.cs

```diff
@@ -0,0 +1,16 @@
+﻿using System;
```

Comment:
object item = Activator.CreateInstance(_collectionValues.GetType().BaseType.GetGenericArguments()[0]);

So I went and changed the collection type to be the concrete type instead of the interface.  I got the above reflection to work.  Notice I had to dive down to the base class of the collection to get the generic arguments.  So this line of code is pretty particular to implementation of the collection.  I am not sure this is any better and pretty fragile and not self documenting.

The other issue we have with this collection is that it has logic in the CanRemove method that does not allow the user to delete the last item in the collection.  I need to be able to delete the last item to create the fixture effect cleanly.  I added logic to allow deleting the last item if the collection item type was known via the IDiscoverCollectionItemType.

How would you like to proceed on these two issues?

Reply:
After looking at this a little more, I am not sure there is an elegant solution. Your original interface based solution is probably better in the long run than the obscure reflection given there are interfaces involved. Are the collections that are being used here generic? If so would it be possible to make this interface take a generic argument that would be the same as the one used when creating the collection. Then there could be an abstract class defines the method which could just return type of T and each collection would not need to implement anything. Might make it a bit cleaner and abstract the actual implementation aside from extending the abstract class.

Reply:
The collections were generic but with the interface type.  I added the concrete impl type as a generic argument as well.  I also created a new common base class for the collections (ExpandoObjectObservableCollection.cs).  I renamed the interface to IExpandoObjectCollection.cs and added a method to this interface that returns the minimum number of items allowed in the collection.  I also split out a new collection class for collections that are dealing with the mark collections.  I updated the effects to use these collections.